# Напишите заголовок проекта здесь

- Автор: Artem Sokolskii
- Дата: 15.04

### Цели и задачи проекта

<font color='#777778'>Используем библиотеку  pandas и датасет `/datasets/new_games.csv`, который содержит информацию о данных компьютерных игр. В датасете познакомимся с данными, проверим их корректность и проведем предобработку, получив необходимый срез данных. </font>

### Описание данных

<font color='#777778'>В проекте будут использованы данные датасета `hotel_dataset.csv` с таким описанием:
- `Name` — название игры;
- `Platform` — название платформы.;
- `Year of Release` — год выпуска игры;
- `Genre` — жанр игры;
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.
</font>

### Содержимое проекта

<font color='#777778'>

1. Загрузка и знакомство с данными
    
2. Проверка ошибок в данных и их предобработка
    
3. Фильтрация данных
    
4. Категоризация данных
</font>

## 1. Загрузка данных и знакомство с ними


In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные из датасета hotel_dataset.csv в датафрейм hotels
new_games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [3]:
# Выводим информацию о датафрейме
new_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Датасет `new_games.csv` содержит 11 столбцов и 16955 строк, в которых представлена информация о данных компьюетрных игр.

Изучим типы данных и их корректность:

- **Числовые значения с плавающей запятой (float64).** Четыре столбца имеют тип данных `float64`: 

    - `Year of Release` - больше подойдет целочисленный тип данных

    - `NA sales` - верное решение, так как количество в миллионах проданных копий может включать дробные значения

    - `Other sales ` - верное решение, так как количество в миллионах проданных копий может включать дробные значения

`Critic Score` - больше подойдет целочисленный тип данных
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `Name`, `Platform` и `Genre` содержат строковую информацию (название игры, платформы и жанр), что логично для текстовых данных. Здесь тип данных `object` подходит.
    - `EU sales` и `JP sales` хранят информацию о продажах (в миллионах проданных копий). Для таких данных рекомендуется использовать тип `float64`, который облегчит выполнение вычислений и упростит анализ.
    - `User Score` и `Rating` также хранят текстовые данные, их необходимо преобразовать в тип данных с плавающей точкой `float64`

После анализа типов данных видно, что в половине столбцов присутствуют пропуски, а в трех из них их около половины, столбцы необходимо преобразовать в более подходящий формат, а также можно использовать целочисленные типы с уменьшенной разрядностью.

In [4]:
# Выводим первые строки датафрейма на экран
new_games.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма


In [5]:
# Выведем все названия столбцов
print(new_games.columns)

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


In [6]:
# Преобразуем столбцы в snake_case (нижний регистр), заменим пробелы на '_'
new_games.columns = (new_games.columns.str.lower().str.replace(' ', '_'))
print(new_games.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


### 2.2. Типы данных

In [7]:
# Приведем 'year_of_release' к типу данных int с понижением размерности значений
new_games['year_of_release'] = new_games['year_of_release'].fillna(0)
new_games['year_of_release'] = pd.to_numeric(new_games['year_of_release'], downcast='integer')


new_games['critic_score'] = new_games['critic_score'].astype('Int16')

In [8]:
# Приведем следующие к типу данных float с понижением размерности значений:
new_games['eu_sales'] = pd.to_numeric(new_games['eu_sales'], errors='coerce', downcast='float')


new_games['jp_sales'] = pd.to_numeric(new_games['jp_sales'], errors='coerce', downcast='float')


new_games['user_score'] = pd.to_numeric(new_games['user_score'], errors = 'coerce', downcast = 'float')

In [9]:
# Снизим размерность:
new_games['na_sales'] = pd.to_numeric(new_games['na_sales'], downcast='float')


new_games['other_sales'] = pd.to_numeric(new_games['other_sales'], downcast='float')

### 2.3. Наличие пропусков в данных


In [10]:
# Выводим количество пропущенных строк в датафрейме
new_games.isna().sum()

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [11]:
# Подсчитываем доли строк с пропусками
new_games.isna().mean()

name               0.000118
platform           0.000000
year_of_release    0.000000
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

В данных наблюдаются пропуски в следующих столбцах:
- `name`: в 2 строках (0.01% данных) отсутствует информация о названии игр. Это может затруднить идентификацию, если необходимо провести анализ по конкретным играм.
- `genre`: в 2 строках (0.01% данных). Пропуски в этом столбце могут повлиять на анализ по жанру игры. Возможно, пропуски в этом поле связаны с пропусками в `name`, учитывая общее количество пропущенных строк.
- `eu_sales`: в 6 строках (0.03% данных) отсутствуют данные о продажах копий в европе.
- `jp_sales`: в 4 строках (0.02% данных) отсутствуют данные о продажах копий в европе.
- `critic_score`: в 8714 строках (51.39% данных) отсутствуют данные о оценках критиков. 
- `user_score`: в 9268 строках (54.65% данных) отсутствуют данные по оценкам пользователей.
- `rating`: в 6871 строках (40.52% данных) отсутствуют данные о рейтинге игр ассоциации ESRB. 

Есть прямая связь между `critic_score`, `user_score`, `rating`, почти во всех этих стоблцах есть пропуски у около половины строк данных, возможно предположить, что эти игры являются старыми и не получали никаких оценок во время выхода игры или не все игры успели получить оценки 

In [12]:
# Удалим все пропуски без названия и жанра
new_games = new_games.dropna(subset=['name', 'genre'])

In [13]:
# Удаляем пропсуки без значений продаж в европе и японии
new_games['eu_sales'] = new_games.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean').fillna(0)
new_games['jp_sales'] = new_games.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean').fillna(0)

In [14]:
# Заполняем пропуски индикаторами
new_games['critic_score'] = new_games['critic_score'].fillna(-1)
new_games['user_score'] = new_games['user_score'].fillna('tbd')
new_games['rating'] = new_games['rating'].fillna('Unknown')

In [15]:
# Проверяем critic_score
print("Минимальное значение critic_score:", new_games['critic_score'].min())

# Проверяем user_score
print(new_games['user_score'].unique()[:10])

# Проверяем rating
print(new_games['rating'].unique())

Минимальное значение critic_score: -1
[8.0 'tbd' 8.300000190734863 8.5 6.599999904632568 8.399999618530273
 8.600000381469727 7.699999809265137 6.300000190734863 7.400000095367432]
['E' 'Unknown' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']


### 2.4. Явные и неявные дубликаты в данных


In [16]:
# Выявим значения по дубликатам
unique_count = new_games['name'].nunique() 
print(f"Количество уникальных строк: {unique_count}")
print(f"Количество строк всего: {len(new_games)}")
range_unique_games = len(new_games) - unique_count
print((f"Количество дубликатов: {range_unique_games}"))

Количество уникальных строк: 11559
Количество строк всего: 16954
Количество дубликатов: 5395


In [17]:
# Изучим уникальные значения в каждом столбце

print("Жанр игры (genre):")
print(new_games['genre'].unique())

print("Платформа (platform):")
print(new_games['platform'].unique())

print("Рейтинг (rating):")
print(new_games['rating'].unique())

print("Год выпуска (year_of_release):")
print(new_games['year_of_release'].unique())

Жанр игры (genre):
['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']
Платформа (platform):
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
Рейтинг (rating):
['E' 'Unknown' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']
Год выпуска (year_of_release):
[2006 1985 2008 2009 1996 1989 1984 2005 1999 2007 2010 2013 2004 1990
 1988 2002 2001 2011 1998 2015 2012 2014 1992 1997 1993 1994 1982 2016
 2003 1986 2000    0 1995 1991 1981 1987 1980 1983]


In [18]:
# Нормализуем данные
new_games['name'] = new_games['name'].str.lower().str.strip()
new_games['platform'] = new_games['platform'].str.upper().str.strip()
new_games['genre'] = new_games['genre'].str.lower().str.strip()
new_games['rating'] = new_games['rating'].str.upper().str.strip()

In [19]:
# Удаляем дубликаты
new_games = new_games.drop_duplicates()

**ПРОМЕЖУТОЧНЫЙ ВЫВОД: ОБРАБОТКА ДУБЛИКАТОВ**

Найденные дубликаты:
- Явные дубликаты: 5395 строк (31.82% от всех данных)
- Неявные дубликаты: проблемы с регистром в столбце 'genre'

Выполненные действия:
- Нормализация текстовых столбцов (приведение к единому регистру, удаление пробелов)
- Удаление всех явных дубликатов методом drop_duplicates(subset='name', keep=False)


**УДАЛЁННЫE СТРОКИ В ПРОЦЕССЕ ПОДГОТОВКИ ДАННЫХ**

Исходные данные:
Начальное количество строк: 16954

Удалённые строки:
  1. Пропуски в name и genre:          2 строк  ( 0.01%)
  2. Явные дубликаты:                5395 строк  (31.82%)

Всего удалено:                     5397 строк  (31.83%)

Результат после обработки:
Конечное количество строк:         11557 строк  (68.17%)

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [20]:
# Отбираем данные по году с 2000 по 2013
new_games = new_games[(new_games['year_of_release'] >= 2000) & (new_games['year_of_release'] <= 2013)]

## 4. Категоризация данных

In [21]:
# Созадем категории для оценок пользователей
def cat_user():
    new_games['user_score_category'] = pd.cut(pd.to_numeric(new_games['user_score'], errors='coerce'), bins=[0,3,8,10], labels=["низкая оценка", "средняя оценка", "высокая оценка"], right=False).cat.add_categories(['нет оценки']).fillna('нет оценки')
    print(new_games['user_score_category'].value_counts())
cat_user()

нет оценки        6298
средняя оценка    4081
высокая оценка    2286
низкая оценка      116
Name: user_score_category, dtype: int64


In [22]:
# Созадем категории для оценок критиков
def cat_critic():
    new_games['critic_score_category'] = pd.cut(pd.to_numeric(new_games['critic_score'], errors='coerce'), bins=[0,30,80,100], labels=["низкая оценка", "средняя оценка", "высокая оценка"], right=False).cat.add_categories(['нет оценки']).fillna('нет оценки')
    print(new_games['critic_score_category'].value_counts())
cat_critic()

нет оценки        5612
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: critic_score_category, dtype: int64


In [23]:
# выводим топ-7 платформ по количеству выпущенных игр
new_games.groupby('platform')['name'].count().sort_values(ascending = False).head(7)

platform
PS2     2127
DS      2120
WII     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: name, dtype: int64

## 5. Итоговый вывод

- Из 16 954 исходных строк после проверки дубликатов по всему датафрейму методом .duplicated().sum() полных дубликатов не обнаружено. Удалены только 2 строки с пропусками в столбцах name и genre (0.01%). После фильтрации данных за период 2000–2013 годов в датафрейме осталась 12 781 строка (75.4% от исходных данных). Проведена нормализация текстовых столбцов (приведение к нижнему регистру, удаление пробелов), заполнены пропуски в продажах средними значениями по платформе и году, а в оценках и рейтингах — индикаторами (-1, 'tbd', 'Unknown').

- Добавлены два новых поля: user_score_category и critic_score_category, которые разделяют игры на категории низкой (0–3 балла), средней (3–8 баллов) и высокой (8–10 баллов) оценки. Выделены топ-7 платформ по количеству игр за исследуемый период (PS2, DS, WII, PSP, X360, PS3, GBA). Всего в обработанном датафрейме 12 781 запись с 13 столбцами, типы данных оптимизированы для экономии памяти.

In [24]:
new_games.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_score_category,critic_score_category
0,wii sports,WII,2006,sports,41.360001,0.991364,0.207955,8.45,76,8.0,E,высокая оценка,средняя оценка
2,mario kart wii,WII,2008,racing,15.680000,0.168556,0.044683,3.29,82,8.3,E,высокая оценка,высокая оценка
3,wii sports resort,WII,2009,sports,15.610000,0.178489,0.050937,2.95,80,8.0,E,высокая оценка,высокая оценка
6,new super mario bros.,DS,2006,platform,11.280000,0.113088,0.189167,2.88,89,8.5,E,высокая оценка,высокая оценка
7,wii play,WII,2006,misc,13.960000,0.991364,0.207955,2.84,58,6.6,E,средняя оценка,средняя оценка


In [25]:
new_games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12781 entries, 0 to 16954
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   name                   12781 non-null  object  
 1   platform               12781 non-null  object  
 2   year_of_release        12781 non-null  int16   
 3   genre                  12781 non-null  object  
 4   na_sales               12781 non-null  float32 
 5   eu_sales               12781 non-null  float32 
 6   jp_sales               12781 non-null  float32 
 7   other_sales            12781 non-null  float32 
 8   critic_score           12781 non-null  Int16   
 9   user_score             12781 non-null  object  
 10  rating                 12781 non-null  object  
 11  user_score_category    12781 non-null  category
 12  critic_score_category  12781 non-null  category
dtypes: Int16(1), category(2), float32(4), int16(1), object(5)
memory usage: 886.6+ KB
